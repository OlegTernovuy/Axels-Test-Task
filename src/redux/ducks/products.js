import { createSlice, createAction } from '@reduxjs/toolkit';
import { put, call, takeEvery } from '@redux-saga/core/effects';

import { getRequest } from '../../api';

const usersInitialState = {
  products: [],
  product: {},
  prodLoading: false,
  singleProdLoading: false,
  errors: '',
};

export const productsSlice = createSlice({
  name: 'products',
  initialState: usersInitialState,
  reducers: {
    getProductsAction: (state, action) => {
      state.errors = '';
      state.prodLoading = true;
    },
    getSingleProductAction: (state, action) => {
      state.errors = '';
      state.singleProdLoading = true;
    },
    getProductSuccessAction: (state, action) => {
      state.products = action.payload;
      state.prodLoading = false;
    },
    getSingleProductSuccessAction: (state, action) => {
      state.product = action.payload;
      state.singleProdLoading = false;
    },
    getProductErrorAction: (state, action) => {
      state.prodLoading = false;
      state.singleProdLoading = false;
      state.errors = action.payload;
    },
  },
});

export const fetchProducts = createAction('products/getProductsAction');

export const fetchSingleProduct = createAction('products/getSingleProductAction');

// Worker
function* getProductsSaga() {
  try {
    const products = yield call(() => getRequest('/prod'));
    yield put(getProductSuccessAction(products));
  } catch (error) {
    yield put(getProductErrorAction(error.message))
  }
}

function* getSingleProductSaga(id) {
  try {
    const product = yield call(() => getRequest(`/prod/${1}`));
    yield put(getSingleProductSuccessAction(product));
  } catch (error) {
    yield put(getProductErrorAction(error.message))
  }
}

// Watcher
export function* watchGetProducts() {
  yield takeEvery(fetchProducts, getProductsSaga);
}

export function* watchGetSingleProduct() {
  yield takeEvery(fetchSingleProduct, getSingleProductSaga);
}

export const {
  getProductsAction,
  getSingleProductAction,
  getProductSuccessAction,
  getSingleProductSuccessAction,
  getProductErrorAction,
} = productsSlice.actions;
export default productsSlice.reducer;
