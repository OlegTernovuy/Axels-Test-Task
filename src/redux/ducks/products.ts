import { createSlice, createAction, PayloadAction } from '@reduxjs/toolkit';
import { put, call, takeEvery } from '@redux-saga/core/effects';
import { type SagaIterator } from '@redux-saga/core';

import { getRequest } from '../../api';
import { IProducts, IProductsState, ISingleProduct } from '../../types';

const usersInitialState: IProductsState = {
  products: [],
  product: {} as ISingleProduct,
  prodLoading: false,
  singleProdLoading: false,
  errors: '',
};

export const products = createSlice({
  name: 'products',
  initialState: usersInitialState,
  reducers: {
    getProductsAction: (state) => {
      state.errors = '';
      state.prodLoading = true;
    },
    getSingleProductAction: (state) => {
      state.errors = '';
      state.singleProdLoading = true;
    },
    getProductSuccessAction: (state, action: PayloadAction<IProducts>) => {
      state.products = action.payload;
      state.prodLoading = false;
    },
    getSingleProductSuccessAction: (state, action: PayloadAction<ISingleProduct>) => {
      state.product = action.payload;
      state.singleProdLoading = false;
    },
    getProductErrorAction: (state, action: PayloadAction<string>) => {
      state.prodLoading = false;
      state.singleProdLoading = false;
      state.errors = action.payload;
    },
  },
});

export const fetchProducts = createAction('Products/getProductsAction');

export const fetchSingleProduct = createAction('Products/getSingleProductAction');

// Worker
function* getProductsSaga(): SagaIterator {
  try {
    const products: IProducts = yield call(() => getRequest<IProducts>('/prod'));    
    yield put(getProductSuccessAction(products));
  } catch (error) {    
    yield put(getProductErrorAction(error.toString()));
  }
}

function* getSingleProductSaga(id): SagaIterator {
  try {
    const product: ISingleProduct = yield call(() => getRequest<ISingleProduct>(`/prod/${1}`));
    yield put(getSingleProductSuccessAction(product));
  } catch (error) {
    yield put(getProductErrorAction(error.toString()))
  }
}

// Watcher
export function* watchGetProducts(): SagaIterator {
  yield takeEvery(fetchProducts, getProductsSaga);
}

export function* watchGetSingleProduct(): SagaIterator {
  yield takeEvery(fetchSingleProduct, getSingleProductSaga);
}

export const {
  getProductsAction,
  getSingleProductAction,
  getProductSuccessAction,
  getSingleProductSuccessAction,
  getProductErrorAction,
} = products.actions;
export default products.reducer;
