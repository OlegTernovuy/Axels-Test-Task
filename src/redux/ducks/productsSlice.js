import { createSlice, createAction } from '@reduxjs/toolkit';
import { put, takeEvery } from '@redux-saga/core/effects';

const usersInitialState = {
  products: [],
  product: {},
  loading: false,
};

export const productsSlice = createSlice({
  name: 'Products',
  initialState: usersInitialState,
  reducers: {
    getProductAction: (state, action) => {
      state.loading = true;
    },
    getProductSuccessAction: (state, action) => {
      state.products = action.payload;
      state.loading = false;
    },
  },
});

export const fetchProducts = createAction('Products/getProductAction');

// Worker
function* getProductSaga() {
  try {
    const response = yield fetch(`http://demo2663713.mockable.io/prod`);
    const json = yield response.json();
    yield put(getProductSuccessAction(json.data));
  } catch (error) {
    console.log(error);
  }
}

// Watcher
export function* watchGetProducts() {
  yield takeEvery(fetchProducts, getProductSaga);
}

export const {
  getProductAction,
  getProductSuccessAction,
  getProductsErrorAction,
} = productsSlice.actions;
export default productsSlice.reducer;
