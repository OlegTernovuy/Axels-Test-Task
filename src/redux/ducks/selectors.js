const products = (store) => store.productsStore.products;
const product = (store) => store.productsStore.product;
const prodLoading = (store) => store.productsStore.prodLoading;
const singleProdLoading = (store) => store.productsStore.singleProdLoading;
const errors = (store) => store.productsStore.errors;

const Selectors = { 
    products,
    product, 
    prodLoading, 
    singleProdLoading, 
    errors
};
export default Selectors;
