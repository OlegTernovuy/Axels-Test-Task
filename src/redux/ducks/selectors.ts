import { RootState } from "..";

const products = (store: RootState) => store.productsStore.products;
const product = (store: RootState) => store.productsStore.product;
const prodLoading = (store: RootState) => store.productsStore.prodLoading;
const singleProdLoading = (store: RootState) => store.productsStore.singleProdLoading;
const errors = (store: RootState) => store.productsStore.errors;

const Selectors = { 
    products,
    product, 
    prodLoading, 
    singleProdLoading, 
    errors
};
export default Selectors;
