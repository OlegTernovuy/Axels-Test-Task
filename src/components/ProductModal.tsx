import { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { Modal, Typography } from '@mui/material';
import Divider from '@mui/material/Divider';

import { CommentsList, CommentForm } from './index';
import {
    BoxModalStyle,
    LoadingSpan,
    ProdInModal,
} from '../styled/modal/StyledModal';
import { ModalCloseButton } from '../styled/modal/StyledCloseModalButton';
import { StyledCardMedia } from '../styled/StyledProductCard';

import { useAppDispatch, useAppSelector } from '../hooks/redux';
import { fetchSingleProduct } from '../redux/ducks/products';
import Selectors from '../redux/ducks/selectors';

const ProductModal = () => {
    const loading = useAppSelector(Selectors.singleProdLoading);
    const singleProduct = useAppSelector(Selectors.product);
    const dispatch = useAppDispatch();
    const navigate = useNavigate();
    const { id } = useParams();

    const [open, setOpen] = useState(true);

    useEffect(() => {
        if (id) dispatch(fetchSingleProduct());
    }, [id, dispatch]);

    const closeModal = () => {
        setOpen(false);
        navigate('/');
    };

    return (
        <div>
            <Modal open={open} onClose={closeModal}>
                <BoxModalStyle>
                    <ModalCloseButton onClick={closeModal} />
                    {loading ? (
                        <LoadingSpan>Loading...</LoadingSpan>
                    ) : (
                        <>
                            <ProdInModal>
                                <StyledCardMedia
                                    height="220"
                                    src={singleProduct.img}
                                    alt="product"
                                />
                                <Typography variant="h5" component="h2">
                                    {singleProduct.title}
                                </Typography>
                                <Typography variant="body2">
                                    {singleProduct.desc}
                                </Typography>
                            </ProdInModal>
                            <Divider />
                            <CommentForm />
                            <Divider />
                            <CommentsList comments={singleProduct.comments} />
                        </>
                    )}
                </BoxModalStyle>
            </Modal>
        </div>
    );
};

export default ProductModal;
