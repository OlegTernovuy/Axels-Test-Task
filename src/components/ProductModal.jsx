import { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { CardMedia, Modal, Typography } from '@mui/material';
import Divider from '@mui/material/Divider';

import { CommentsList, CommentForm } from './index';
import { BoxModalStyle, ProdInModal } from '../styled/modal/StyledModal';
import { ModalCloseButton } from '../styled/modal/StyledCloseModalButton';

import { fetchSingleProduct } from '../redux/ducks/productsSlice';
import Selectors from '../redux/selectors';

const ProductModal = () => {
  const dispatch = useDispatch()
  const navigate = useNavigate();
  const { prodId } = useParams();

  const [open, setOpen] = useState(true);
  const singleProduct = useSelector(Selectors.product);
  const loading = useSelector(Selectors.singleProdLoading);

  useEffect(() => {
    if (!prodId) return undefined;
    dispatch(fetchSingleProduct());
  }, [prodId, dispatch]);

  const closeModal = () => {
    setOpen(false);
    navigate('/');
  };

  return (
    <Modal open={open} onClose={closeModal}>
      <BoxModalStyle>
        <ModalCloseButton onClick={closeModal} />
        {loading ? (
          <>Loading...</>
        ) : (
          <>
            <ProdInModal>
              <CardMedia
                component='img'
                height='220'
                image={singleProduct.img}
                alt='product'
              />
              <Typography variant='h5' component='h2'>
                {singleProduct.title}
              </Typography>
              <Typography variant='body2'>{singleProduct.desc}</Typography>
            </ProdInModal>
            <Divider />
            <CommentForm />
            <Divider />
            <CommentsList comments={singleProduct.comments} />
          </>
        )}
      </BoxModalStyle>
    </Modal>
  );
};

export default ProductModal;
