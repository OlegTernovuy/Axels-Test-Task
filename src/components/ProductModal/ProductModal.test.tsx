import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import ProductModal from './ProductModal';
import { fetchSingleProduct } from '../../redux/ducks/products';

const mockUseParams = jest.fn();
const mockedUsedNavigate = jest.fn();
const mockUseDispatch = jest.fn();
const mockUseSelector = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useParams: () => mockUseParams(),
    useNavigate: () => mockedUsedNavigate,
}));

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux'),
    useDispatch: () => mockUseDispatch,
    useSelector: () => mockUseSelector,
}));

jest.mock('../../redux/ducks/products', () => ({
    fetchSingleProduct: jest.fn(),
}));

describe('Product modal component', () => {
    beforeEach(() => {
        mockUseParams.mockReturnValue({ id: '1' });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('Matches Snapshot', () => {
        const { asFragment } = render(<ProductModal />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('Dispatches fetchSingleProduct on mount', () => {
        render(<ProductModal />);
        expect(mockUseDispatch).toHaveBeenCalledWith(fetchSingleProduct());
    });

    it('renders loading state correctly', () => {
        render(<ProductModal />);
        expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
    });
});
