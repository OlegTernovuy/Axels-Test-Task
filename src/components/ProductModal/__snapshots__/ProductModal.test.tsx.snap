// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product modal component Matches Snapshot 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;
